[inductor][dynamo] Include operator name in size/stride/alignment assertion

karthickai opened the pull request:
Fixes #151930 

This PR updates the `assert_size_stride` and `assert_alignment` functions in [guards.cpp](https://github.com/pytorch/pytorch/blob/main/torch/csrc/dynamo/guards.cpp) to accept an optional `op_name` argument and includes it in the error messages.

The corresponding type stubs in [guards.pyi](https://github.com/pytorch/pytorch/blob/main/torch/_C/_dynamo/guards.pyi) are updated to match the new function arg.

In [inductor/ir.py](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py) extracts the operator name from the FX graph and passes it into the `codegen_size_asserts` and `codegen_alignment_asserts` functions, so that generated assertions in Triton code include the op name for better debugging.

Added unit tests inside [test_torchinductor.py](https://github.com/pytorch/pytorch/blob/main/test/inductor/test_torchinductor.py).
- Verified both successful and failing assertion cases include the operator name.
- Verified that generated Triton code contains the op name inside the asserts.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @shunting314 @eellison

karthickai committed:
[dynamo] Added optional op_name argument to assert_size_stride and assert_alignment functions

karthickai committed:
[dynamo] Add assert_alignment stub with op_name and update assert_size_stride stub with op_name in guards.pyi

karthickai committed:
[inductor] Pass operator name into codegen_size_asserts and codegen_alignment_asserts

karthickai committed:
[inductor][test] Add tests for assert_size_stride and assert_alignment with operator name

karthickai committed:
[inductor][test] Add missing import for assert_size_stride and assert_alignment

karthickai commented:
@pytorchbot label "topic: not user facing"

jansel commented on torch/_inductor/ir.py:
```diff
@@ -5713,25 +5713,37 @@ def codegen_kwargs(self, skip_out=False):  # type: ignore[no-untyped-def]
+            op_name = self.get_op_name()          
             wrapper.writeline(
-                f"assert_size_stride({self.get_name()}, {size}, {stride})"
+                f"assert_size_stride({self.get_name()}, {size}, {stride}, '{op_name}')"
```
```suggestion
                f"assert_size_stride({self.get_name()}, {size}, {stride}, {op_name!r})"
```

jansel commented on torch/_inductor/ir.py:
```diff
@@ -5713,25 +5713,37 @@ def codegen_kwargs(self, skip_out=False):  # type: ignore[no-untyped-def]
         return kwargs
 
+    def get_op_name(self) -> str:
+        if hasattr(self.fx_node, "target"):
```
Is this only happening when fx_node is None?  My understanding is target is a required field.   If so let's check for None rather than hasattr.

karthickai replied:
Switched to `self.fx_node is not None` per your suggestion. Appreciate the catch!

jansel commented on torch/_inductor/ir.py:
```diff
@@ -5713,25 +5713,37 @@ def codegen_kwargs(self, skip_out=False):  # type: ignore[no-untyped-def]
+            op_name = self.get_op_name() 
             if aligned:
-                wrapper.writeline(f"assert_alignment({name}, {GPU_ALIGN_BYTES})")
+                wrapper.writeline(f"assert_alignment({name}, {GPU_ALIGN_BYTES}, '{op_name}')")
```
```suggestion
                wrapper.writeline(f"assert_alignment({name}, {GPU_ALIGN_BYTES}, {op_name!r})")
```

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -896,23 +916,35 @@ static PyObject* assert_size_stride(PyObject* dummy, PyObject* args) {
   Py_RETURN_TRUE;
 }
-
+ 
```
Remove whitespace.

jansel commented:
Also looks like lints are failing.

karthickai committed:
[inductor] Address reviewer feedback: clean up formatting, fix mypy issues, and simplify op_name handling

karthickai commented:
Thank you @jansel, I’ve pushed updates with your suggestions.

shunting314 commented:
The lint job still fails. Not sure if this is some transient error though

jansel commented:
You should be able to fix the lints with:
```
make setup_lint
lintrunner -a
```

karthickai committed:
[inductor][test] Fix lint issues in test_torchinductor.py

karthickai commented:
@jansel @shunting314 I believe the lint error is now resolved with the latest commit.

jansel commented:
@pytorchbot merge

karthickai commented:
@pytorchbot label "skip-url-lint"

karthickai committed:
[inductor][test] Remove dynamic shape in test cases, update existing test, and fix op name construction

karthickai commented:
cc @jansel @shunting314. I've added the skip-url-lint label to bypass transient outages in the Lint link check. In the latest commit, I updated existing tests to reflect recent changes in `assert_size_stride` and fixed the `get_op_name` to avoid duplicate namespace segments when combining op_namespace and target.

jansel reviewed (changes requested):
Test failures?

karthickai commented:
cc @jansel, I looked into the failed test cases: most failures are from`inductor/test_torchinductor.py::test_add_complex4`. I believe this PR doesn’t touch any logic related to complex ops, so these failures are likely unrelated. The remaining failures are from `inductor/test_mkldnn_pattern_matcher.py::test_woq_int4`, which asserts on the number of included ops. Since this PR introduces op names in the asserts, they’re now being counted, so this failure appears to be an expected side effect.

jansel commented:
If the failures are unrelated you could try rebasing onto the `viable/strict` branch (which is updated automatically when tests pass).  We should update any remaining tests.

karthickai committed:
[inductor] [test] Fix two test failures by filtering ops in codegen checks and matching appropriate op counts

karthickai commented:
cc @jansel  In the latest commit, I fixed both test cases. It looks like test_complex_add4 was also failing due to an op count mismatch, not unrelated as I originally thought. I resolved it by updating the expected op count to reflect the new behavior.

jansel commented:
@pytorchbot merge

karthickai committed:
[inductor] [test] Fix test_add_complex4 op_count to match cpp_wrapper

karthickai commented:
@pytorchbot rebase -b viable/strict

jansel commented:
@pytorchbot rebase -b viable/strict

jansel commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but it seems to fail an inductor test in trunk' -c nosignal

inductor/test_torchinductor.py::GPUTests::test_generated_code_has_size_stride_assert_cuda [GH job link](https://github.com/pytorch/pytorch/actions/runs/14902124829/job/41859344117) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/93576351270383ca37deaec6b2417a33dc045a93)

karthickai committed:
[inductor][test] Fix test_generated_code by adding skip_cpp_wrapper annotation

karthickai commented:
@pytorchbot label ciflow/inductor

karthickai commented:
@pytorchbot label ciflow/trunk

karthickai commented:
@pytorchbot rebase -b viable/strict

karthickai commented:
@pytorchbot label ciflow/inductor

karthickai commented:
@pytorchbot rebase -b viable/strict

karthickai commented:
@pytorchbot label ciflow/inductor

karthickai commented:
@pytorchbot label ciflow/inductor

jansel commented:
@pytorchbot merge

jansel commented:
@pytorchbot merge

jansel commented:
@pytorchbot merge -i

karthickai commented:
@pytorchbot label ciflow/inductor ciflow/trunk

karthickai committed:
[inductor] [test] Fix op count in distributed test_all_to_all_single_inductor

karthickai commented:
@pytorchbot label ciflow/inductor ciflow/trunk

karthickai commented:
@pytorchbot label ciflow/trunk

jansel commented:
@pytorchbot merge

jansel commented:
@pytorchbot merge -i

jeanschmidt commented:
@pytorchbot revert -m "seems to have broken a few internal tests, @jansel may you help the author get his PR merged?" -c ghfirst

jeanschmidt commented:
Forgot to share internal ref: [D74804066](https://www.internalfb.com/diff/D74804066)

karthickai commented:
@jansel @shunting314, could you please share details of failed internal tests?

jansel commented:
Failing tests are:
```
caffe2/test/inductor:compile_subprocess - test_add_complex4_cuda (caffe2.test.inductor.test_compile_subprocess.GPUTests)

caffe2/test/inductor:compile_subprocess - test_generated_code_has_alignment_assert_cuda (caffe2.test.inductor.test_compile_subprocess.GPUTests) 

caffe2/test/inductor:compile_subprocess - test_remove_noop_slice1_cuda (caffe2.test.inductor.test_compile_subprocess.GPUTests) 

caffe2/test/inductor:compile_subprocess - test_remove_noop_slice_cuda (caffe2.test.inductor.test_compile_subprocess.GPUTests) 

caffe2/test/inductor:compile_subprocess - test_remove_noop_slice_scatter_cuda (caffe2.test.inductor.test_compile_subprocess.GPUTests) 

caffe2/test/inductor:compile_subprocess - test_remove_noop_view_default_cuda (caffe2.test.inductor.test_compile_subprocess.GPUTests) 

caffe2/test/inductor:compile_subprocess - test_remove_noop_view_dtype_cuda (caffe2.test.inductor.test_compile_subprocess.GPUTests) 

caffe2/test/inductor:test_inductor_cuda - test_add_complex4_cuda (caffe2.test.inductor.test_torchinductor.GPUTests) 

caffe2/test/inductor:test_inductor_cuda - test_generated_code_has_alignment_assert_cuda (caffe2.test.inductor.test_torchinductor.GPUTests) 

caffe2/test/inductor:torchinductor_codegen_dynamic_shapes - test_add_complex4_dynamic_shapes_cuda (caffe2.test.inductor.test_torchinductor_codegen_dynamic_shapes.DynamicShapesCodegenGPUTests) 

caffe2/test/inductor:torchinductor_dynamic_shapes - test_add_complex4_dynamic_shapes_cuda (caffe2.test.inductor.test_torchinductor_dynamic_shapes.DynamicShapesGPUTests)
```
These are all public tests (ignore the caffe2 prefix, this is a fbcode artifact).  Not sure why they are not failing in open source CI.

Some example failures:
```py
self.assertMultiLineEqualMaybeCppStack(expect, actual, msg=help_text)
  File "/re_cwd/buck-out/v2/gen/fbcode/a1488de98b7ab8de/caffe2/test/inductor/__compile_subprocess__/compile_subprocess#link-tree/expecttest/__init__.py", line 292, in assertMultiLineEqualMaybeCppStack
    self.assertMultiLineEqual(expect, actual, *args, **kwargs)
AssertionError: 'def forward(self, arg0_1: "Sym(s77)", arg[333 chars]_9,)' != ''
- def forward(self, arg0_1: "Sym(s77)", arg1_1: "Sym(s27)", arg2_1: "Sym(s53)", arg3_1: "f32[s77, s27, s53][s27*s53, s53, 1]cuda:0"):
-         add: "f32[s77, s27, s53][s27*s53, s53, 1]cuda:0" = torch.ops.aten.add.Tensor(arg3_1, 1);  arg3_1 = None
-         add_9: "f32[s77, s27, s53][s27*s53, s53, 1]cuda:0" = torch.ops.aten.add.Tensor(add, 1);  add = None
-         return (add_9,) : To accept the new output, re-run test with envvar EXPECTTEST_ACCEPT=1 (we recommend staging/committing your changes before doing this)
```

```py
FAIL: test_add_complex4_cuda (caffe2.test.inductor.test_torchinductor.GPUTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/re_cwd/buck-out/v2/gen/fbcode/a1488de98b7ab8de/caffe2/test/inductor/__test_inductor_cuda__/test_inductor_cuda#link-tree/caffe2/test/inductor/test_torchinductor.py", line 13488, in new_test
    return value(self)
  File "/re_cwd/buck-out/v2/gen/fbcode/a1488de98b7ab8de/caffe2/test/inductor/__test_inductor_cuda__/test_inductor_cuda#link-tree/caffe2/test/inductor/test_torchinductor.py", line 1417, in test_add_complex4
    self.assertEqual(code.count("aten.view"), 9)
  File "/re_cwd/buck-out/v2/gen/fbcode/a1488de98b7ab8de/caffe2/test/inductor/__test_inductor_cuda__/test_inductor_cuda#link-tree/torch/testing/_internal/common_utils.py", line 4089, in assertEqual
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: Scalars are not equal!

Expected 9 but got 6.
Absolute difference: 3
Relative difference: 0.3333333333333333
```

karthickai commented:
@pytorchbot rebase

karthickai commented:
@pytorchbot label ciflow/inductor ciflow/trunk

karthickai committed:
[inductor] [test] Fix test cases using regex and add filtered asserts in generated code

karthickai committed:
[inductor] [test] Fix test_add_complex4 by using assertGreaterEqual on op count

karthickai commented:
@pytorchbot label ciflow/inductor ciflow/trunk

karthickai committed:
[inductor] [test] Add debug log and relax regex in test_generated_code_has_size_stride_assert and test_generated_code_has_alignment_assert

jansel commented:
@pytorchbot merge -f "landed in fbcode"
